README
 * hcrond - a cron with crontab in mysql
 *
 *  Copyright 2006, 2007, Martin Hradil <himdel@seznam.cz>
 *  Licensed under GPL the terms of which can be found in the file COPYING.

 * table:
 * id = unique identifier
 * sec = 0..59
 * min = 0..59
 * hour = 0..23
 * day = 1..31
 * mon = 1..12
 * dow = 0..6 (sunday = 0, 7)
 * uid = EUID for cmd
 * gid = EGID for cmd
 * machine = hostname where to run (everywhere when null)
 * cmd = cmd for sh -c
 * name = task name
 * andor = ('&', '|') specifies whether to AND or OR the dow with day&mon
 * nice = nice value
 * lastrun = 0 (used to record time cmd was last run) (is INT!)
 * runonce = we run the job only N times and then delete it (-1 = infinite)
 *
 * standard cron syntax is supported, so "1-5/2" == "1,3,5" = 
 *  = "Mon,Wed,Fri" or "Jan,Mar,May" and * is every
 *
 * TODO negative numbers signify frequency (except for sec) (min == -2 => run twice per min)

Usage: hcrond [OPTION]...
hcrond is a cron with crontab in mysql

Mandatory arguments to long options are mandatory for short options too.
Bool with no args defaults to true (true, 1, yes)
  -c, --configfile	configfile filename (/etc/hcrondrc)
  -P, --pidfile		pidfile (/var/run/hcrond.pid)  ## not implemented
  -h, --host		address (IP or hostname) of server running mysql (127.0.0.1)
  -o, --port		the port mysql listens to (666)
  -u, --user		username (evil)
  -p, --pass		password (swordfish)
  -d, --dbnm		database name (hcrond)
  -t, --table		table name (crontab)
  -R, --allow_root	allow running cmd with uid=0 or gid=0 (bool:no)  ## not implemented
  -g, --allow_uidgid	allow running cmd with different uid && gid (bool:yes)  ## not implemented
  -U, --force_uid	run everything with this uid or username  ## not implemented
  -G, --force_gid	run everything with this gid or groupname  ## not implemented
  -H, --force_hostname	use this hostname for comparison with `machine' field  ## not implemented
  -i, --ignore_machine	ignore the `machine' field (bool:no)  ## not implemented - always true
  -n, --allow_notnice	allow negative nice values (bool:no) ## not implemented
  -s, --force_shell	always use this shell  ## not implemented
  -D, --force_shell_die	die if forced shell different from `sh' (bool:yes)  ## not implemented
  -j, --max-jobs	run only n jobs at a time; 0 means infinity (1)
  -r, --reload		reload from table every n secs (30)
  -e, --debug		debug mode (no detach and more verbosity) (bool:no)
  --help		display this help and exit
  -V, --version		output version information and exit

Report bugs to <himdel@seznam.cz>.

sql to create table:

CREATE TABLE `hcrondtab` (
`id` INT NOT NULL AUTO_INCREMENT PRIMARY KEY ,
`sec` VARCHAR( 128 ) NULL DEFAULT '*',
`min` VARCHAR( 128 ) NULL DEFAULT '*',
`hour` VARCHAR( 128 ) NULL DEFAULT '*',
`day` VARCHAR( 128 ) NULL DEFAULT '*',
`mon` VARCHAR( 128 ) NULL DEFAULT '*',
`dow` VARCHAR( 128 ) NULL DEFAULT '*',
`uid` VARCHAR( 128 ) NULL ,
`gid` VARCHAR( 128 ) NULL ,
`machine` VARCHAR( 128 ) NULL ,
`cmd` VARCHAR( 512 ) NOT NULL ,
`name` VARCHAR( 128 ) NOT NULL ,
`andor` ENUM( '&', '|' ) NOT NULL DEFAULT '|',
`nice` INT NULL DEFAULT '0',
`lastrun` INT NULL DEFAULT '0',
`runonce` INT DEFAULT '-1' NOT NULL
) ENGINE = MYISAM ;

Thanks to:
	moonlake.cz		for sponsoring this project
	libdaemon authors	for libdaemon
	sourceforge.net		for hosting both hcrond and libdaemon


ChangeLog
hcrond-0.1.1 (released 2011-01-18)
 * bugfix:
  * no jobs got run in January if mon="\*" and andor="&"
  * also mon="1" meant February, not January
  * thanks to Philipp Schreiber for reporting it
 * more TODO

hcrond-0.4.0 (released 2007-01-01)
 * pidfile support set as compile-time option => config.h
 + tasks are organized in a priority queue => performance, ...

hcrond-0.3.2-nopid (released 2006-12-30)
 * pidfile support removed

hcrond-0.3.2 (released 2006-12-29)
 * logging macros now really use daemon_log

hcrond-0.3.1 (released 2006-12-28)
 + uid, gid, nice, allow_uidgid, allow_notnice

hcrond-0.3 (released 2006-12-28)
 * lots of changes

hcrond-0.2 (released 2006-08-24)
 * we now depend on libdaemon
 + pidfile implemented
 * logging macros (dbg, err, inf, wrn) now use daemonlog
 * macro scp(d, s) replaced by d = strdup(s)
 + add option -k, --kill to kill the running hcrond
 + SIGHUP, SIGINT and SIGQUIT handled correctly
 	INT & QUIT update the db and die
	HUP update the db and start anew, remembering jobs_running
 * removed locking as none needed
 * no need for garbage list, removed
 * job list is now just a single linked list (should become a tree one day)
 * thanks in README
 * trans now handles NULL inputs (only the 3rd param can be NULL)
 * split database refresh stuff
 * signals are handled by libdaemon, in run_jobs and main
 	libdaemon provides daemon_signal_next
	run_jobs handles SIGALRM and SIGCHLD, other signals mean return
	main handles SIGHUP and dies on SIGINT and SIGQUIT
 * TODO TODO

hcrond-0.1 (~ Aug 8, 2006)
 * initial release (except for various unstable -rc's)
 * uses --std=gnu99 because of sigaction
 * more TODO
 * uses sigaction, doesn't do SIGHUP
 * some dbg stuff commented out, why?

hcrond 0.1-rc3 (~ Aug 7, 2006)
 * locking fixed - phthread & semaphore
 * pidfile
 * more debug info

hcrond 0.1-rc1 (~ Jun 31, 2006)
 * has license gpl2
 * empty changelog
 * makefile using flex and libmysqlclient
 * readme with opts and db table
 * todo list with some bugs
 * hcrond sysv initscript
 * random locking


help.c
/* help.c for hcrond */
/* when updating remember to update README too */

void
help()
{
	printf("Usage: hcrond [OPTION]...\n"
		"hcrond is a cron with crontab in mysql\n\n"
		"Mandatory arguments to long options are mandatory for short options too.\n"
		"Bool with no args defaults to true (true, 1, yes)\n"
		"  -c, --configfile\tconfigfile filename (/etc/hcrondrc)\n"
		"  -P, --pidfile\t\tpidfile (/var/run/hcrond.pid)  ## not implemented\n"
		"  -h, --host\t\taddress (IP or hostname) of server running mysql (127.0.0.1)\n"
		"  -o, --port\t\tthe port mysql listens to (666)\n"
		"  -u, --user\t\tusername (evil)\n"
		"  -p, --pass\t\tpassword (swordfish)\n"
		"  -d, --dbnm\t\tdatabase name (hcrond)\n"
		"  -t, --table\t\ttable name (crontab)\n"
		"  -R, --allow_root\tallow running cmd with uid=0 or gid=0 (bool:no)  ## not implemented\n"
		"  -g, --allow_uidgid\tallow running cmd with different uid && gid (bool:yes)  ## not implemented\n"
		"  -U, --force_uid\trun everything with this uid or username  ## not implemented\n"
		"  -G, --force_gid\trun everything with this gid or groupname  ## not implemented\n"
		"  -H, --force_hostname\tuse this hostname for comparison with `machine' field  ## not implemented\n"
		"  -i, --ignore_machine\tignore the `machine' field (bool:no)  ## not implemented - always true\n"
		"  -n, --allow_notnice\tallow negative nice values (bool:no) ## not implemented\n"
		"  -s, --force_shell\talways use this shell  ## not implemented\n"
		"  -D, --force_shell_die\tdie if forced shell different from `sh' (bool:yes)  ## not implemented\n"
		"  -r, --reload\t\treload from table every n secs (30)\n"
		"  -j, --max-jobs\t\trun only n jobs at a time; 0 means infinity (1)\n"
		"  -e, --debug\t\tdebug mode (no detach and more verbosity) (bool:no)\n"
		"  -k, --kill\t\tkill already running hcrond\n"
		"  --help\t\tdisplay this help and exit\n"
		"  -V, --version\t\toutput version information and exit\n\n"
		"Report bugs to <himdel@seznam.cz>.\n");
	exit(0);
}
